packages/web/src/common/store/upload/sagas.js

```diff
@@ -898,262 +902,92 @@ function* uploadCollection(tracks, userId, collectionMetadata, isAlbum) {
-            uploadActions.singleTrackUploadError(
-              message,
-              phaseContainer.phase,
-              track.file.size
-            )
-          )
         }
-        yield cancel(dispatcher)
-        yield call(responseChan.put, { error: message })
-      },
-      () => {},
-      UPLOAD_TIMEOUT_MILLIS
+      })
     )
   )
 
-  const { confirmedTrack, error } = yield take(responseChan)
-  const isRejected = error === 'Request failed with status code 403'
+  // Get stem metadatas
+  const stems = yield select(getStems)
+  const updatedStems = updateAndFlattenStems(
+    stems,
+    tracksWithMetadata.map((t) => t.metadata.id)
+  )
+  const allUploads = tracksWithMetadata.concat(updatedStems)
 
-  yield reportResultEvents({
-    numSuccess: error ? 0 : 1,
-    numFailure: error && !isRejected ? 1 : 0,
-    numRejected: isRejected ? 1 : 0,
-    uploadType: 'single_track',
-    errors: error ? [error] : []
+  // Upload tracks and stems parallel together
+  const { trackIds } = yield call(handleUploads, {
+    tracks: allUploads,
+    isCollection: false
   })
 
-  if (error) {
-    if (isRejected) {
-      yield put(
-        make(Name.TRACK_UPLOAD_COMPLETE_UPLOAD, {
-          count: 1,
-          kind: 'tracks'
-        })
+  // Record stem upload events
+  for (let i = tracksWithMetadata.length; i < trackIds.length; i += 1) {
+    const trackId = trackIds[i]
+    const parentTrackId = allUploads[i]?.metadata?.stem_of?.parent_track_id
+    const category = allUploads[i]?.metadata?.stem_of?.category
+    if (!trackIds.includes(parentTrackId)) {
+      console.debug(
+        `Stem ${trackId} (${allUploads[i].metadata.title}) was not successful as its parent wasn't uploaded`
       )
-      yield call(recordGatedTracks, [track.metadata])
+      continue
     }
-    return
-  }
-
-  const stems = yield select(getStems)
-  if (stems.length) {
-    yield call(uploadStems, {
-      parentTrackIds: [confirmedTrack.track_id],
-      stems
+    const recordEvent = make(Name.STEM_COMPLETE_UPLOAD, {
+      id: trackId,
+      parent_track_id: parentTrackId,
+      category
     })
+    yield put(recordEvent)
   }
 
-  // Finish up the upload
-  progressChan.put(
-    uploadActions.updateProgress(0, 'art', {
-      status: ProgressStatus.COMPLETE
-    })
-  )
-  progressChan.put(
-    uploadActions.updateProgress(0, 'audio', {
-      status: ProgressStatus.COMPLETE
-    })
-  )
-
-  yield put(
-    uploadActions.uploadTracksSucceeded(confirmedTrack.track_id, [
-      confirmedTrack
-    ])
-  )
-  yield put(
-    make(Name.TRACK_UPLOAD_COMPLETE_UPLOAD, {
-      count: 1,
-      kind: 'tracks'
-    })
-  )
-  yield call(recordGatedTracks, [confirmedTrack])
+  // If EVERYTHING failed, don't mark this as a success!
+  if (trackIds.length === 0 || trackIds.every((t) => t === undefined)) {
+    yield put(uploadActions.uploadTrackFailed())
+    yield put(
+      toastActions.toast({ content: 'Something went wrong.', type: 'error' })
```

Comment:
what happens if this toasts? do we just stay on the progress bar page? that seems not great b/c the user may be on another tab and miss it? I think the error page makes more sense here

Reply:
no strong feelings here, just felt it's easier to try again w/ clicking upload than taken super far out of the app experience, and figured we were trying to move away from that page and towards toasts. also wasn't sure of mobile behavior (still need to test that)

Reply:
can you click upload when you get into this though? i think it would show a spinner

Reply:
yeah it's always there on the sidebar - the action above fails the upload (not sure what happens on mobile though tbh, testing)

Reply:
I think if you're already on the upload page though and it fails, it wouldn't be obvious to click that again would it?

Reply:
Not qualified to answer that haha, but changed to put handleError and redirect regardless

I think the ideal solution here would be to have some similar hero to the success one rather than a redirect but that's beyond the scope here

Reply:
Yes 100% agreed. I think we could do a lot more to map upload errors. There are some that are also fatal and will fail on retry, e.g. 429